### PR TITLE
Set up project with 3 routes

### PR DESCRIPTION
**Pull request by vitokhangnguyen**
Hi,
I have the functionality of login, register and check existence done. You can take a look and tell me if you want to change anything 😄 
I will also make the client-side tmessage interact with this API.

### CODE REVIEW

**Comment by Haider8** [comment]
@vitokhangnguyen Great work!

I am getting this error for the route `localhost:8080/api/user/register` :
```
SyntaxError: Unexpected token ' in JSON at position 3
    at JSON.parse (<anonymous>)
    at parse (/home/haider/github-repos/tmessage-api/node_modules/body-parser/lib/types/json.js:89:19)
    at /home/haider/github-repos/tmessage-api/node_modules/body-parser/lib/read.js:121:18
    at invokeCallback (/home/haider/github-repos/tmessage-api/node_modules/raw-body/index.js:224:16)
    at done (/home/haider/github-repos/tmessage-api/node_modules/raw-body/index.js:213:7)
    at IncomingMessage.onEnd (/home/haider/github-repos/tmessage-api/node_modules/raw-body/index.js:273:7)
    at IncomingMessage.emit (events.js:188:13)
    at endReadableNT (_stream_readable.js:1129:12)
    at process.internalTickCallback (internal/process/next_tick.js:72:19)
```

**Comment by vitokhangnguyen** [comment]
Hi @Haider8,
I am not having that error on my side, can you check 2 things?
* You did `npm i` to install all the packages
* Your JSON sent in the body of the request is well-formatted, for eg:
```json
{
	"userName": "vitokhangnguyen",
	"displayedName": "Khang Nguyen",
	"password": "password",
	"passwordConfirm": "password"
}
```

**Comment by Haider8** [comment]
Now I am getting this response
```
{
    "success": false,
    "message": "not authorized on admin to execute command { insert: \"userAccount\", documents: [[{_id ObjectIdHex(\"5d95807a651f3e1e88e91997\")} {userName vitokhangnguyen} {displayedName Khang Nguyen} {password $2a$10$JBbQOnJq.7KfYe/QbbsiBeDz42nO2ZQ3ymNaXkFqIjGF4AhHBaOxO} {__v 0}]], ordered: true, writeConcern: { w: \"majority\" }, lsid: { id: {4 [246 245 25 101 102 131 64 67 168 212 198 114 160 25 213 177]} }, txnNumber: 1.000000, $clusterTime: { clusterTime: 6743437274236584936, signature: { hash: [207 182 28 93 27 160 154 69 160 16 144 62 135 242 161 92 116 188 209 43], keyId: 6740875669841903616.000000 } }, $db: \"admin\" }"
}
```

**Comment by vitokhangnguyen** [comment]
I suspect that error rises from the setting in the MongoDB. If you are using MongoDB Atlas, check the Database Access panel:
image
Make sure you have an admin user there and substitute the credential into the connection string.
```
mongodb+srv://[ADMIN_USERNAME]:[ADMIN_PASSOWRD]@bti425-cluster-jyzoe.azure.mongodb.net/[DATABASE_NAME]?retryWrites=true&w=majority
```

**Comment by Haider8** [comment]
Yes, I checked that. Can you please share the format of your connection string here?

**Comment by vitokhangnguyen** [comment]
Here it is:
```
mongodb+srv://[ADMIN_USERNAME]:[ADMIN_PASSOWRD]@bti425-cluster-jyzoe.azure.mongodb.net/[DATABASE_NAME]?retryWrites=true&w=majority
```

**Comment by vitokhangnguyen** [comment]
There's one more thing I would syggest you to check. Make sure your cluster allows connection from all IP (0.0.0.0).

**Comment by Haider8** [comment]
Thanks for your reply. I have done that already. I think there's some issue
with my connection string.

**Comment by vitokhangnguyen** [comment]
I did some Googling and maybe you can try...
* Use connection string for an earlier version of Nodejs (you can get this in MongoDB Atlas)
* Create a collection on MongoDB Atlas before running the API

I hope this helps...

File: server.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+// Security with JWT
+const jwt = require('jsonwebtoken');
+const passport = require("passport");
+const passportJWT = require("passport-jwt");
+const ExtractJwt = passportJWT.ExtractJwt;
+const JwtStrategy = passportJWT.Strategy;
+
+var jwtOptions = {};
+jwtOptions.jwtFromRequest = ExtractJwt.fromAuthHeaderAsBearerToken();
+jwtOptions.secretOrKey = 'xIyC#bzK$42AmKfU8GyHw^9sPtF&C2$Ic$mKolkiU6XVK88KYG';
```
Comment (iamareebjamal) [inline server.js]: Make this an environment variable

File: server.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+passport.use(strategy);
+app.use(passport.initialize());
+
+// Add this middleware to routes need protected
+const basicTokenAuthentication = passport.authenticate("jwt", {session: false});
+
+app.post('/api/user/register', async (req, res) => {
+    try {
+        let user = await auth.register(req.body);
```
Comment (iamareebjamal) [inline server.js:49]: There should be data validation that user has sent data in correct format?

Maybe using joi or alternative https://github.com/hapijs/joi or https://docs.nestjs.com/techniques/validation

File: auth-manager.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                db.on('error', err => reject(err));
+                db.once('open', _ => {
+                    UserAccount = db.model('userAccount', userAccountSchema, 'userAccount');
+                    resolve(db);
+                });
+            });
+        },
+
+        register: async function(newUser) {
+            if (!newUser.userName || newUser.userName == '') {
```
Comment (iamareebjamal) [inline auth-manager.js]: Use underscores as separators instead of camel case. It makes it extremely difficult to create clients in mobile

**Comment by Haider8** [comment]
> Here it is:
> 
> ```
> mongodb+srv://[ADMIN_USERNAME]:[ADMIN_PASSOWRD]@bti425-cluster-jyzoe.azure.mongodb.net/[DATABASE_NAME]?retryWrites=true&w=majority
> ```

This works now @vitokhangnguyen . MongoDB was generating the wrong connection URL for me. Thank you!

**Review by iamareebjamal — COMMENTED** [review]
I know this is very nascent stage, but perhaps it should be using a proper framework like nestjs to avoid common problems https://docs.nestjs.com/techniques/mongodb

Also, I recommend typeorm so that if in future, the requirement is to use a relational database(as one should), it'll be easy to integrate https://github.com/typeorm/typeorm

Also, will make it easy to swap for sqlite for local development

**Comment by Haider8** [comment]
@vitokhangnguyen All the routes are working fine.

Work on the changes suggested by @iamareebjamal and we can then merge this PR.

File: server.js (hunk 1 above)
Comment (Haider8) [reply server.js]: Do this change

**Comment by iamareebjamal** [comment]
You can choose between https://foalts.org/ and nest later on, but at least change camel case to underscores and move secret key to env

File: auth-manager.js (hunk 3 above)
Comment (Haider8) [reply auth-manager.js]: And do this change also. The rest of the changes can be done later. I will open new issues for that.

**Comment by vitokhangnguyen** [comment]
Hi,
I will make the change soon. Thank you for all of your effort and feedbacks.

**Comment by Haider8** [comment]
@vitokhangnguyen Just do those two changes. We can do other changes later on.

**Comment by vitokhangnguyen** [comment]
I've made the 2 changes. I will work on the client-side now to enable user authentication.

**Review by iamareebjamal — APPROVED** [review]

**Pull request merged**